File: MainModule/src/main/java/com/qihuan/find/presenter/DailyPresenter.java
package com.qihuan.find.presenter;

import com.qihuan.commonmodule.utils.DateUtils;
import com.qihuan.find.contact.DailyContract;
import com.qihuan.find.model.bean.zhihu.DailyBean;
import com.qihuan.find.model.bean.zhihu.DailyItemBean;
import com.qihuan.find.model.remote.ZhihuModel;
import com.qihuan.find.presenter.base.BasePresenterImpl;

import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class DailyPresenter extends BasePresenterImpl<DailyContract.View> implements DailyContract.Presenter {

    private final ZhihuModel zhihuModel;

    public DailyPresenter() {
        zhihuModel = new ZhihuModel();
    }

    @Override
    public void getLatestDaily() {
        checkViewAttached();
        getView().showLoading();
        disposables.add(
            zhihuModel.getApi().getLatestDaily()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnNext(dailyBean -> getView().latestDaily(dailyBean.getTop_stories()))
                .observeOn(Schedulers.io())
                .concatMap(dailyBean -> Flowable.fromIterable(dailyBean.getStories()))
//                        .flatMap(storyBean -> Flowable.zip(Flowable.just(storyBean), zhihuApi.getStoryExtra(storyBean.getId()), StoryBean::setStoryExtraBean))
                .map(DailyItemBean::new)
                .toList()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    list -> {
                        list.add(0, new DailyItemBean(true, "Toady"));
                        getView().beforeDaily(list);
                        getView().onRefreshEnd();
                    },
                    e -> getView().showError(e.getMessage())
                )
        );
    }

    @Override
    public void getBeforeDaily(String date) {
        checkViewAttached();
        final DailyBean[] d = new DailyBean[1];
        disposables.add(
            zhihuModel.getApi().getBeforeDaily(date)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnNext(dailyBean -> d[0] = dailyBean)
                .observeOn(Schedulers.io())
                .concatMap(dailyBean -> Flowable.fromIterable(dailyBean.getStories()))
//                        .flatMap(storyBean -> Flowable.zip(Flowable.just(storyBean), zhihuApi.getStoryExtra(storyBean.getId()), StoryBean::setStoryExtraBean))
                .map(DailyItemBean::new)
                .toList()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    list -> {
                        list.add(0, new DailyItemBean(true, DateUtils.parseDate(d[0].getDate())));
                        getView().beforeDaily(list);
                        getView().onLoadMoreEnd(true);
                    },
                    e -> {
                        getView().onLoadMoreEnd(false);
                        getView().showError(e.getMessage());
                    }
                )
        );
    }
}

File: module_daily/src/main/java/com/qihuan/dailymodule/presenter/DailyPresenter.kt
package com.qihuan.dailymodule.presenter

import com.qihuan.commonmodule.base.AbsRxPresenter
import com.qihuan.commonmodule.utils.getNowDate
import com.qihuan.commonmodule.utils.parseDate
import com.qihuan.commonmodule.utils.timeSub
import com.qihuan.dailymodule.contract.DailyContract
import com.qihuan.dailymodule.model.ZhihuApi
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.rxkotlin.subscribeBy
import io.reactivex.schedulers.Schedulers

/**
 * DailyPresenter
 *
 * @author Qi
 */
class DailyPresenter : AbsRxPresenter<DailyContract.View>(), DailyContract.Presenter {

    private var date: String = getNowDate()

    override fun getLatestDaily() {
        view?.showLoading()
        date = getNowDate()

        ZhihuApi.get()
                .getLatestDaily()
                .doOnSubscribe { addDisposable(it) }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(
                        onNext = {
                            view?.onRefreshStart()
                            view?.onBannerData(it.top_stories)
                            view?.onDailyData(it.stories)
                            view?.onRefreshEnd(true)
                        },
                        onError = {
                            view?.onRefreshEnd(false)
                            view?.showError(it.message ?: "")
                        }
                )
    }

    override fun getBeforeDaily() {
        date = timeSub(date)
        ZhihuApi.get()
                .getBeforeDaily(date)
                .doOnSubscribe { addDisposable(it) }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(
                        onNext = {
                            view?.onDailySectionData(parseDate(date))
                            view?.onDailyData(it.stories)
                            view?.onLoadMoreEnd(true)
                        },
                        onError = {
                            view?.onLoadMoreEnd(false)
                            view?.showError(it.message ?: "")
                        }
                )
    }
}
